### ISSUE TITLE
TypeInferrer doesn't propogate through method type arguments

### ISSUE THREAD

**Issue by dpoeschl** [dotnet/roslyn#4233]
From customer:
```C#
using System.Threading.Tasks;

public class C
{
    private async Task M()
    {
        float[] result = await Task.Run(() => Test());
    }
}
```
**Expected** Generate Method on ```Test``` should generate a return type of float[]
**Actual** It generates a ```void``` method.

Another repro, with no overloads on the invoked method:
```C#
using System.Threading.Tasks;

public class C
{
    private async void M()
    {
        float[] merged = await X(Test());
    }

    private async Task<T> X<T>(T t) { return t; }
}
```
**Expected** Generate Method on ```Test``` should generate a return type of ```float[]```
**Actual** It generates a return type of ```object```

**Comment by dpoeschl** [comment]
The TypeInferrer propogates the inferred type of the await expression to the return type of the method being directly invoked, so it knows Task.Run should return a Task<float[]>. It doesn't currently take the next step of realizing that the signature of Run (Task<TResult> Run<TResult>(Func<TResult> function)) means that the lambda should be a Func<float[]> in this case.

**Comment by CyrusNajmabadi** [comment]
Fixed with 9d6aef71a6ff7a5b8e8ab57efac26b7e1cd82a2a
